pkg/xds/topology/outbound.go

```diff
@@ -2,38 +2,62 @@ package topology
 
 import (
 	"context"
-	"net"
 
 	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
 	mesh_core "github.com/Kong/kuma/pkg/core/resources/apis/mesh"
 	core_manager "github.com/Kong/kuma/pkg/core/resources/manager"
 	core_store "github.com/Kong/kuma/pkg/core/resources/store"
+	core_xds "github.com/Kong/kuma/pkg/core/xds"
 )
 
-func GetOutboundTargets(ctx context.Context, dataplane *mesh_core.DataplaneResource, manager core_manager.ResourceManager) (map[string][]net.SRV, error) {
-	outbound := make(map[string][]net.SRV)
-	if len(dataplane.Spec.Networking.GetOutbound()) > 0 {
-		dataplanes := &mesh_core.DataplaneResourceList{}
-		if err := manager.List(ctx, dataplanes, core_store.ListByMesh(dataplane.Meta.GetMesh())); err != nil {
-			return nil, err
-		}
-		for _, oface := range dataplane.Spec.Networking.GetOutbound() {
-			outbound[oface.Service] = make([]net.SRV, 0)
-		}
-		for _, dataplane := range dataplanes.Items {
-			for _, inbound := range dataplane.Spec.Networking.GetInbound() {
-				service := inbound.Tags[mesh_proto.ServiceTag]
-				endpoints, ok := outbound[service]
-				if !ok {
-					continue
-				}
-				iface, err := mesh_proto.ParseInboundInterface(inbound.Interface)
-				if err != nil {
-					return nil, err
+// GetOutboundTargets resolves all endpoints reachable from a given dataplane.
+func GetOutboundTargets(ctx context.Context, dataplane *mesh_core.DataplaneResource, destinations core_xds.DestinationMap, manager core_manager.ResourceManager) (core_xds.EndpointMap, error) {
+	if len(destinations) == 0 {
+		return nil, nil
+	}
+	dataplanes := &mesh_core.DataplaneResourceList{}
+	if err := manager.List(ctx, dataplanes, core_store.ListByMesh(dataplane.Meta.GetMesh())); err != nil {
+		return nil, err
+	}
+	return BuildEndpointMap(destinations, dataplanes.Items), nil
+}
+
+// BuildEndpointMap creates a map of all endpoints that match given selectors.
+func BuildEndpointMap(destinations core_xds.DestinationMap, dataplanes []*mesh_core.DataplaneResource) core_xds.EndpointMap {
+	if len(destinations) == 0 {
+		return nil
+	}
+	outbound := core_xds.EndpointMap{}
+	for _, dataplane := range dataplanes {
+		for _, inbound := range dataplane.Spec.Networking.GetInbound() {
+			service := inbound.Tags[mesh_proto.ServiceTag]
+			selectors, ok := destinations[service]
+			if !ok {
+				continue
+			}
+			matches := false
+			for _, selector := range selectors {
+				if selector.Matches(inbound.Tags) {
+					matches = true
+					break
 				}
-				outbound[service] = append(endpoints, net.SRV{Target: iface.DataplaneIP, Port: uint16(iface.DataplanePort)})
 			}
+			if !matches {
+				continue
+			}
+			iface, err := mesh_proto.ParseInboundInterface(inbound.Interface)
+			if err != nil {
+				// skip dataplanes with invalid configuration
+				continue
+			}
+			// TODO(yskopets): do we need to dedup?
```

Comment:
I guess it would be safer to do it. Can you maybe add an issue so we don't forget about it?

Reply:
https://github.com/Kong/kuma/issues/427